zeth-contracts/contracts/Groth16Mixer.sol

```diff
@@ -43,31 +83,274 @@ contract Groth16Mixer is BaseMixer {
         add_and_emit_merkle_root(getRoot());
 
-        // Emit the all the coins' secret data encrypted with the recipients'
+        // 6. Emit the all the coins' secret data encrypted with the recipients'
         // respective keys
         emit_ciphertexts(pk_sender, ciphertext0, ciphertext1);
     }
+
+    function verify(
+        uint[] memory input,
+        Proof memory proof) internal returns (uint) {
+
+        // `input.length` = size of the instance = l (see notations in the
+        // reference paper).  We have coefficients indexed in the range[1..l],
+        // where l is the instance size, and we define a_0 = 1. This is the
+        // reason we need to check that: input.length + 1 == vk.ABC.length (the
+        // +1 accounts for a_0). This equality is a strong consistency check
+        // (len(givenInputs) needs to equal expectedInputSize (not less))
+        require(
+            input.length + 1 == verifyKey.ABC.length,
+            "Input length differs from expected");
+
+        // A memory scratch pad
+        uint[24] memory pad;
+
+        // 1. Compute the linear combination
+        //   vk_x = \sum_{i=0}^{l} a_i * vk.ABC[i], vk_x in G1.
+        //
+        // ORIGINAL CODE:
+        //   Pairing.G1Point memory vk_x = vk.ABC[0]; // a_0 = 1
+        //   for (uint i = 0; i < input.length; i++) {
+        //       vk_x = Pairing.add(vk_x, Pairing.mul(vk.ABC[i + 1], input[i]));
+        //   }
+        //
+        // The linear combination loop was the biggest cost center of the mixer
+        // contract.  The following assembly block removes a lot of unnecessary
+        // memory usage and data copying, but relies on the structure of storage
+        // data.
+        //
+        // `pad` is layed out as follows, (so that calls to precompiled
+        // contracts can be done with minimal data copying)
+        //
+        //  OFFSET  USAGE
+        //   0x20    accum_y
+        //   0x00    accum_x
+
+        // In each iteration, copy scalar multiplicaation data to 0x40+
+        //
+        //  OFFSET  USAGE
+        //   0x80    input_i   --
+        //   0x60    abc_y      | compute abc[i+1] * input[i] in-place
+        //   0x40    abc_x     --
+        //   0x20    accum_y
+        //   0x00    accum_x
+        //
+        //  ready to call bn256ScalarMul(in: 0x40, out: 0x40).  This results in:
+        //
+        //  OFFSET  USAGE
+        //   0x80
+        //   0x60    input_i * abc_y  --
+        //   0x40    input_i * abc_x   |  accum = accum + input[i] * abc[i+1]
+        //   0x20    accum_y           |
+        //   0x00    accum_x          --
+        //
+        //  ready to call bn256Add(in: 0x00, out: 0x00) to update accum_x,
+        //  accum_y in place.
+
+        bool success = true;
+        assembly {
+
+            let g := sub(gas, 2000)
+
+            // Compute slot of ABC[0], using pad[0] as a temporary
+            mstore(pad, add(verifyKey_slot, 10))
```

Comment:
Where is `verifyKey_slot` declared?

Reply:
Added a comment (it's defined by soildity assembly notation).

Reply:
Ah yes of course, I forgot about this `_slot`/`_offset` thing! I thought it was an undeclared variable. Definitely good to have that documented in the code; it's clearer. Thanks!